File: backend/app/utils/security.py
from werkzeug.security import generate_password_hash, check_password_hash
import re
import bleach
from datetime import datetime, timedelta
from flask import jsonify, current_app

def sanitize_input(input_str, max_length=255):
    """清理和验证用户输入"""
    if not input_str:
        return None
    
    # 去除前后空白
    input_str = input_str.strip()
    
    # 限制长度
    if len(input_str) > max_length:
        raise ValueError(f"输入长度超过最大限制 {max_length} 字符")
    
    # 移除危险字符
    input_str = re.sub(r'[<>"\'%;()&]', '', input_str)
    
    return input_str

def validate_email(email):
    """验证邮箱格式"""
    if not email:
        return False
    
    email = email.strip().lower()
    email_regex = re.compile(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$')
    return bool(email_regex.match(email))

def validate_username(username):
    """验证用户名格式"""
    if not username:
        return False
    
    username = username.strip()
    username_regex = re.compile(r'^[a-zA-Z0-9_]{3,50}$')
    return bool(username_regex.match(username))

def validate_password(password):
    """验证密码强度"""
    if not password:
        return False
    
    password = password.strip()
    if len(password) < 8:
        return False
    
    # 检查密码复杂度
    has_upper = any(c.isupper() for c in password)
    has_lower = any(c.islower() for c in password)
    has_digit = any(c.isdigit() for c in password)
    has_special = any(not c.isalnum() for c in password)
    
    return has_upper and has_lower and has_digit and has_special

def sanitize_search_keyword(keyword):
    """清理搜索关键词，防止SQL注入"""
    if not keyword:
        return None
    
    # 移除危险字符
    keyword = re.sub(r'[<>"\'%;()&]', '', keyword)
    
    # 限制长度
    if len(keyword) > 100:
        raise ValueError("搜索关键词过长")
    
    return f"%{keyword}%"

def sanitize_sort_field(sort_field):
    """验证和清理排序字段"""
    allowed_fields = ['created_at', 'price_asc', 'price_desc', 'student_count']
    if sort_field not in allowed_fields:
        return 'created_at'
    return sort_field

def handle_api_error(error, status_code=400):
    """统一处理API错误"""
    current_app.logger.error(f"API Error: {str(error)}")
    return jsonify({
        'error': str(error),
        'status': 'error'
    }), status_code

def sanitize_filename(filename):
    """清理文件名，防止路径遍历"""
    if not filename:
        return None
    
    # 使用werkzeug的secure_filename
    filename = secure_filename(filename)
    
    # 限制文件名长度
    if len(filename) > 255:
        raise ValueError("文件名过长")
    
    return filename

def validate_file_upload(file, allowed_extensions, max_size):
    """验证上传的文件"""
    if not file:
        raise ValueError("未提供文件")
    
    # 检查文件扩展名
    if '.' not in file.filename:
        raise ValueError("无效的文件扩展名")
    
    extension = file.filename.rsplit('.', 1)[1].lower()
    if extension not in allowed_extensions:
        raise ValueError(f"不支持的文件类型: {extension}")
    
    # 检查文件大小
    file.seek(0, 2)  # 移动到文件末尾
    file_size = file.tell()
    file.seek(0)  # 重置文件指针
    
    if file_size > max_size:
        raise ValueError(f"文件大小超过限制: {max_size / (1024 * 1024):.1f}MB")
    
    return True

def generate_safe_filename(original_filename):
    """生成安全的文件名"""
    if not original_filename:
        return None
    
    # 获取当前时间戳
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    
    # 获取文件扩展名
    if '.' in original_filename:
        extension = original_filename.rsplit('.', 1)[1]
        return f"{timestamp}_{secure_filename(original_filename.replace('.' + extension, ''))}.{extension}"
    else:
        return f"{timestamp}_{secure_filename(original_filename)}"

def hash_password(password):
    """安全地哈希密码"""
    return generate_password_hash(password, method='pbkdf2:sha256')

def check_password(hashed_password, password):
    """验证密码"""
    return check_password_hash(hashed_password, password)

def sanitize_html(content):
    """清理HTML内容，防止XSS攻击"""
    if not content:
        return None
    
    # 使用bleach清理HTML
    allowed_tags = ['p', 'br', 'strong', 'em', 'u', 'ol', 'ul', 'li', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6']
    allowed_attrs = {}
    
    return bleach.clean(content, tags=allowed_tags, attributes=allowed_attrs)

File: frontend/src/utils/security.js
import { ElMessage } from 'element-plus'

// 验证用户名格式
export const validateUsername = (username) => {
  if (!username) {
    return '请输入用户名'
  }
  
  if (username.length < 3 || username.length > 50) {
    return '用户名长度应在3-50个字符之间'
  }
  
  const usernameRegex = /^[a-zA-Z0-9_]+$/
  if (!usernameRegex.test(username)) {
    return '用户名只能包含字母、数字和下划线'
  }
  
  return null
}

// 验证密码强度
export const validatePassword = (password) => {
  if (!password) {
    return '请输入密码'
  }
  
  if (password.length < 6) {
    return '密码长度至少为6个字符'
  }
  
  return null
}

// 验证邮箱格式
export const validateEmail = (email) => {
  if (!email) {
    return '请输入邮箱'
  }
  
  const emailRegex = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/
  if (!emailRegex.test(email)) {
    return '请输入有效的邮箱地址'
  }
  
  return null
}

// 统一处理API错误
export const handleApiError = (error) => {
  console.error('API Error:', error)
  
  if (error.response) {
    const status = error.response.status
    const data = error.response.data
    
    if (status === 400) {
      ElMessage.error(data.error || '请求参数错误')
    } else if (status === 401) {
      ElMessage.error('登录已过期，请重新登录')
    } else if (status === 403) {
      ElMessage.error('无权限访问')
    } else if (status === 404) {
      ElMessage.error('资源未找到')
    } else if (status >= 500) {
      ElMessage.error('服务器错误，请稍后重试')
    } else {
      ElMessage.error(data.error || '操作失败')
    }
  } else if (error.request) {
    ElMessage.error('网络连接失败，请检查网络')
  } else {
    ElMessage.error('未知错误')
  }
  
  return null
}

// 清理输入内容
export const sanitizeInput = (input) => {
  if (!input) return ''
  
  // 移除前后空白
  let sanitized = input.trim()
  
  // 移除危险字符
  sanitized = sanitized.replace(/[<>"'%;()&]/g, '')
  
  return sanitized
}

// 验证文件上传
export const validateFileUpload = (file, maxSize = 50 * 1024 * 1024, allowedTypes = ['image/jpeg', 'image/png', 'image/gif', 'video/mp4']) => {
  if (!file) {
    throw new Error('请选择文件')
  }
  
  if (file.size > maxSize) {
    throw new Error(`文件大小不能超过 ${(maxSize / (1024 * 1024)).toFixed(1)}MB`)
  }
  
  if (!allowedTypes.includes(file.type)) {
    throw new Error('不支持的文件类型')
  }
  
  return true
}